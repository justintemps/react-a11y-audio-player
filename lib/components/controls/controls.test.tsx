import { fireEvent, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import renderer from "react-test-renderer";
import { Pause, Play, ProgressBar, SkipButton } from ".";

function testButtonClick(
  Button: React.FC<{ onClick: () => void }>,
  testId: string
) {
  const cb = jest.fn();
  const { getByTestId } = render(<Button onClick={cb} />);
  const button = getByTestId(testId);
  userEvent.click(button);
  expect(cb).toHaveBeenCalled();
}

function testSkipButtonClick(direction: "forward" | "rewind") {
  const increment = 20;
  const expected = direction === "forward" ? increment : -1 * increment;
  const cb = jest.fn((increment) => increment);
  const { getByTestId } = render(
    <SkipButton direction={direction} handleSkip={cb} increment={increment} />
  );
  const button = getByTestId(`rk-skip-button-${direction}`);
  userEvent.click(button);
  expect(cb.mock.calls[0][0]).toBe(expected);
}

describe("Play button", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Play onClick={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("correctly handles click", () => {
    testButtonClick(Play, "rk-play-button");
  });
});

describe("Pause button", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Pause onClick={() => {}} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("correctly handles click", () => {
    testButtonClick(Pause, "rk-pause-button");
  });
});

describe("Skip button", () => {
  it("renders correctly as fast forward", () => {
    const tree = renderer
      .create(<SkipButton direction="forward" handleSkip={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly as rewind", () => {
    const tree = renderer
      .create(<SkipButton direction="rewind" handleSkip={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes positive int to click handler", () => {
    testSkipButtonClick("forward");
  });

  it("passes negative int to click handler", () => {
    testSkipButtonClick("rewind");
  });
});

describe("ProgressBar", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(<ProgressBar progress={50} handleSetProgress={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("passes percent to click handler", () => {
    const WIDTH = 500;
    const CLICK_LOCATION = 250;

    // Set up our mocked useRef
    const ref = { current: {} };
    Object.defineProperty(ref, "current", {
      set(_current) {
        if (_current) {
          jest
            .spyOn(_current, "getBoundingClientRect")
            .mockReturnValueOnce({ width: WIDTH });
        }
        this._current = _current;
      },
      get() {
        return this._current;
      },
    });

    // Hijack useRef to use our ref as the return value for useRef
    jest.spyOn(React, "useRef").mockReturnValueOnce(ref);

    // Our test callback should return percent
    const cb = jest.fn((percent) => percent);

    // Render the progress bar and get a reference
    const { getByTestId } = render(
      <ProgressBar progress={0} handleSetProgress={cb} />
    );
    const progressBar = getByTestId("rk-progress-bar");

    // Set up the event which should fire when the progress bar gets clicked
    const event = new MouseEvent("click", {
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(event, "offsetX", { get: () => CLICK_LOCATION });

    // Click on the progress bar
    fireEvent(progressBar, event);

    // Expect the callback to receive a percentage of the click location
    // with respect to the width of the progress bar
    expect(cb.mock.calls[0][0]).toBe(CLICK_LOCATION / WIDTH);
  });
});
