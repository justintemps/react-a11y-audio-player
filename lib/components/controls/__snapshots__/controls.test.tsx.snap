// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pause button renders correctly 1`] = `
<button
  aria-label="Pause"
  className="pause"
  data-testid="rk-pause-button"
  onClick={[Function]}
>
  <svg
    viewBox="0 0 40 40"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <circle
        cx="20"
        cy="20"
        fill="black"
        r="19"
      />
      <path
        d="M13 12h4v16h-4zm10 0h4v16h-4z"
        fill="white"
      />
    </g>
  </svg>
</button>
`;

exports[`Play button renders correctly 1`] = `
<button
  aria-label="Play"
  className="play"
  data-testid="rk-play-button"
  disabled={false}
  onClick={[Function]}
>
  <svg
    viewBox="0 0 40 40"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g>
      <circle
        cx="20"
        cy="20"
        fill="black"
        r="19"
      />
      <path
        d="M31.334 19.667l-19.333 9.666V10z"
        fill="white"
      />
    </g>
  </svg>
</button>
`;

exports[`ProgressBar renders correctly 1`] = `
<div
  className="progress_bar_wrapper"
>
  <div
    className="progress_bar"
    data-testid="rk-progress-bar"
    onClick={[Function]}
    style={
      Object {
        "--progress": "50",
      }
    }
  />
</div>
`;

exports[`Skip button renders correctly as fast forward 1`] = `
<button
  aria-label="Forward 10 seconds"
  className="skip_button"
  data-testid="rk-skip-button-forward"
  onClick={[Function]}
>
  <svg
    preserveAspectRatio="xMidYMid meet"
    version="1.1"
    viewBox="0 0 200 200"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="100"
      cy="100"
      fill="white"
      r="98"
      stroke="black"
      strokeWidth="4"
    />
    <text
      style={
        Object {
          "font": "115px monospace",
        }
      }
      textLength="170"
      x="16"
      y="135"
    >
      +10
    </text>
  </svg>
</button>
`;

exports[`Skip button renders correctly as rewind 1`] = `
<button
  aria-label="Rewind 10 seconds"
  className="skip_button"
  data-testid="rk-skip-button-rewind"
  onClick={[Function]}
>
  <svg
    preserveAspectRatio="xMidYMid meet"
    version="1.1"
    viewBox="0 0 200 200"
    xmlns="http://www.w3.org/2000/svg"
  >
    <circle
      cx="100"
      cy="100"
      fill="white"
      r="98"
      stroke="black"
      strokeWidth="4"
    />
    <text
      style={
        Object {
          "font": "115px monospace",
        }
      }
      textLength="170"
      x="16"
      y="135"
    >
      -10
    </text>
  </svg>
</button>
`;
